clang/lib/Sema/SemaExpr.cpp

```diff
@@ -223,9 +223,12 @@ bool Sema::DiagnoseUseOfDecl(NamedDecl *D, ArrayRef<SourceLocation> Locs,
       if (IsRuntimeEvaluated && !IsConst && VD->getStorageClass() == SC_Static)
         SYCLDiagIfDeviceCode(*Locs.begin(), diag::err_sycl_restrict)
             << Sema::KernelNonConstStaticDataVariable;
-      // Non-const globals are allowed for SYCL explicit SIMD.
+      // Non-const globals are allowed for SYCL explicit SIMD or with the
+      // SYCLGlobalVar attribute.
       else if (IsRuntimeEvaluated && !IsEsimdPrivateGlobal && !IsConst &&
-               VD->hasGlobalStorage())
+               VD->hasGlobalStorage() &&
+               !(VD->hasAttr<SYCLGlobalVarAttr>() &&
+                 getSourceManager().isInSystemHeader(*Locs.begin())))
```

Comment:
Once you have the code in place in SemaDeclAttr.cpp to only apply the attribute if it's used within a system header, this check for a system header can go away entirely because there's no way to write the attribute except within a system header.